From 0e04b93ee329e2af291f412af52a0b91890fe081 Mon Sep 17 00:00:00 2001
From: Si Zhiying <sizhiying@loongson.cn>
Date: Thu, 1 Dec 2016 09:51:35 +0800
Subject: [PATCH 130/136] Ugly workarounds for Loongson-3

1, Remove redundant sync in llsc;
2, Use synci as smp_llsc_mb() for new Loongson-3.
---
 arch/mips/include/asm/barrier.h | 5 +++++
 arch/mips/loongson64/Platform   | 1 +
 2 files changed, 6 insertions(+)

diff --git a/arch/mips/include/asm/barrier.h b/arch/mips/include/asm/barrier.h
index c020876f9..979953c50 100644
--- a/arch/mips/include/asm/barrier.h
+++ b/arch/mips/include/asm/barrier.h
@@ -227,6 +227,11 @@
 #define __LLSC_CLOBBER		"memory"
 #endif
 
+#ifdef CONFIG_LOONGSON3_ENHANCEMENT
+#undef  smp_llsc_mb
+#define smp_llsc_mb()	__asm__ __volatile__(".set mips64r2\n synci 0\n .set mips0\n" : : :"memory")
+#endif
+
 #ifdef CONFIG_CPU_CAVIUM_OCTEON
 #define smp_mb__before_llsc() smp_wmb()
 #define __smp_mb__before_llsc() __smp_wmb()
diff --git a/arch/mips/loongson64/Platform b/arch/mips/loongson64/Platform
index 9f79908f5..d4edbd1d6 100644
--- a/arch/mips/loongson64/Platform
+++ b/arch/mips/loongson64/Platform
@@ -45,6 +45,7 @@ cflags-$(CONFIG_CPU_LOONGSON3)	+= -Wa,--trap
 # this later - for now it ensures vendor toolchains don't cause problems.
 #
 cflags-$(CONFIG_CPU_LOONGSON3)	+= $(call as-option,-Wa$(comma)-mno-fix-loongson3-llsc,)
+cflags-$(CONFIG_CPU_LOONGSON3)	+= $(call as-option,-Wa$(comma)-mno-fix-ls3a-llsc,)
 
 #
 # binutils from v2.25 on and gcc starting from v4.9.0 treat -march=loongson3a
-- 
2.39.1

